rfcs/react-components/convergence/slot-render-as.md

```diff
@@ -0,0 +1,195 @@
+
+### Implementation
+
+The `WithSlotRenderFunction` type would be expanded to apply the render function support to the `as` prop:
+
+```ts
+type WithSlotRenderFunction<Props> = Props & {
+  as?: (Props extends { as?: string } ? Props['as'] : never) | SlotRenderFunction<Props>;
+
+  // Existing [deprecated] support for a render function on children:
+  children?: (Props extends { children?: unknown } ? Props['children'] : never) | SlotRenderFunction<Props>;
+};
+```
+
+The `getSlot` function would support `as` being a function:
+
+```ts
+if (typeof asProp === 'function') {
+  const render = asProp as SlotRenderFunction<R[K]>;
+  return [
+    React.Fragment,
+    {
+      children: render(slot, propsWithoutAs),
+    } as unknown as R[K],
+  ];
+}
+```
+
+### Adoption Plan
+
+A render function on the `as` prop is a new feature, and does not require any changes to existing code. We can add this feature without needing to change existing uses of `children` as a render function.
+
+We will continue to support `children` as a render function, with a `console.warning` about deprecation. It will be removed in the next major release.
+
+## Appendix A: [REJECTED] Change the order of arguments
+
+> ❌ This Appendix was rejected due to the possibility of accidentally passing a FunctionComponent as the render function (which could break the rules of hooks).
+
+As mentioned in the problem statement, the order of arguments to the render function could be swapped for improved ergonomics:
+
+- Current: `(DefaultComponent, props)`
+- Proposed: `(props, DefaultComponent)`
+
+This makes it simpler to write a render function that ignores and replaces the component, since it can simply omit the second argument.
+
+For example, before this RFC, a render function might look like:
+
+```jsx
+<Field
+  label={{
+    // Note the first argument being the component type, is being ignored here
+    children: (_ignored: unknown, props: LabelProps) => (
+      <InfoLabel {...props} info="Example info">
+        Field with an info button
+      </InfoLabel>
+    ),
+  }}
+/>
+```
+
+With the proposal in this RFC, it would be:
+
+```jsx
+<Field
+  label={{
+    as: (props: LabelProps) => <InfoLabel {...props} info="Example info" />,
+    children: 'Field with an info button',
+  }}
+/>
+```
+
+## Appendix B: `render` prop instead of `as`
+
+If we wanted to keep the `as` prop scoped to just the DOM element that a slot is rendered as, we could optionally add a new special prop `render`, which has a single purpose of being the render function.
+
+If we do choose to go with a `render` prop, then I'd propse following the argument order from Appendix A: `(props, DefaultComponent)`. It would have the benefits listed in Appendix A, with a lower likelihood that someone would pass in a FunctionComponent by mistake, because the prop is named `render`.
+
+### Pros
+
+1. Doesn't overload the `as` prop with multiple purposes.
+
+### Cons
+
+1. Adds another "special" prop into the slots API.
```

Comment:
I see this as a pro actually, it's a well stablished property on the community to be used as a render function alternative.